engines/supernova/supernova.cpp

```diff
@@ -0,0 +1,1152 @@
+	_soundMusic[1] = convertToMod("msn_data.052");
+
+	// Cursor
+	const uint16 *bufferNormal = reinterpret_cast<const uint16 *>(mouseNormal);
+	const uint16 *bufferWait = reinterpret_cast<const uint16 *>(mouseWait);
+	for (uint i = 0; i < sizeof(mouseNormal) / 4; ++i) {
+		for (uint bit = 0; bit < 16; ++bit) {
+			uint mask = 0x8000 >> bit;
+			uint bitIndex = i * 16 + bit;
+
+			_mouseNormal[bitIndex] = (READ_LE_UINT16(bufferNormal + i) & mask) ? kColorCursorTransparent : kColorBlack;
+			if (READ_LE_UINT16(bufferNormal + i + 16) & mask)
+				_mouseNormal[bitIndex] = kColorLightRed;
+			_mouseWait[bitIndex] = (READ_LE_UINT16(bufferWait + i) & mask) ? kColorCursorTransparent : kColorBlack;
+			if (READ_LE_UINT16(bufferWait + i + 16) & mask)
+				_mouseWait[bitIndex] = kColorLightRed;
+		}
+	}
+}
+
+void SupernovaEngine::initPalette() {
+	_system->getPaletteManager()->setPalette(initVGAPalette, 0, 256);
+}
+
+void SupernovaEngine::playSound(AudioIndex sample) {
+	if (sample > kAudioNumSamples - 1)
+		return;
+
+	Audio::SeekableAudioStream *audioStream = Audio::makeRawStream(
+	            _soundSamples[sample]._buffer, _soundSamples[sample]._length,
+	            11931, Audio::FLAG_UNSIGNED | Audio::FLAG_LITTLE_ENDIAN, DisposeAfterUse::NO);
+	stopSound();
+	_mixer->playStream(Audio::Mixer::kPlainSoundType, &_soundHandle, audioStream);
+}
+
+void SupernovaEngine::stopSound() {
+	if (_mixer->isSoundHandleActive(_soundHandle))
+		_mixer->stopHandle(_soundHandle);
+}
+
+void SupernovaEngine::playSoundMod(int filenumber)
+{
+	if (filenumber != 49 && filenumber != 52) {
+		return;
+	}
+
+	int index = filenumber == 49 ? 0 : 1;
+	Audio::AudioStream *audioStream = Audio::makeProtrackerStream(_soundMusic[index]);
+	stopSound();
+	_mixer->playStream(Audio::Mixer::kMusicSoundType, &_soundHandle, audioStream,
+	                   -1, Audio::Mixer::kMaxChannelVolume, 0, DisposeAfterUse::NO);
+}
+
+void SupernovaEngine::renderImageSection(int section) {
+	// Note: inverting means we are removing the section. So we should get the rect for that
+	// section but draw the background (section 0) instead.
+	bool invert = false;
+	if (section > 128) {
+		section -= 128;
+		invert = true;
+	}
+	if (!_currentImage || section > _currentImage->_numSections - 1)
+		return;
+
+	Common::Rect sectionRect(_currentImage->_section[section].x1,
+	                         _currentImage->_section[section].y1,
+	                         _currentImage->_section[section].x2 + 1,
+	                         _currentImage->_section[section].y2 + 1) ;
+	if (_currentImage->_filenumber == 1 || _currentImage->_filenumber == 2) {
+		sectionRect.setWidth(640);
+		sectionRect.setHeight(480);
+		if (_screenWidth != 640) {
+			_screenWidth = 640;
+			_screenHeight = 480;
+			initGraphics(_screenWidth, _screenHeight);
+		}
+	} else {
+		if (_screenWidth != 320) {
+			_screenWidth = 320;
+			_screenHeight = 200;
+			initGraphics(_screenWidth, _screenHeight);
```

Comment:
Since the engine seems to switch modes between 320x200 and 640x480 perhaps it should declare to the backend the list of used modes using `initGraphicsModes` so the backend can choose a suited window size.